databuilder/sqlalchemy_utils.py

```diff
@@ -0,0 +1,207 @@
+
+    # Find all the selected column names
+    column_names = [column.name for column in query.selected_columns]
+
+    # Query to select the columns that we need to sort on
+    order_columns = [table_expr.c[column] for column in sort_columns]
+    # change ordering to descending on all order columns if necessary
+    if descending:
+        order_columns = [c.desc() for c in order_columns]
+
+    # Number rows sequentially over the order by columns for each patient id
+    row_num = (
+        sqlalchemy.func.row_number()
+        .over(order_by=order_columns, partition_by=table_expr.c[partition_column])
+        .label("_row_num")
+    )
+    # Add the _row_num column and select just the first row
+    query = query.add_columns(row_num)
+    subquery = query.alias()
+    query = sqlalchemy.select([subquery.c[column] for column in column_names])
+    query = query.select_from(subquery).where(subquery.c._row_num == 1)
+    return query
+
+
+def group_and_aggregate(
+    query, group_by_column, input_column, function_name, output_column
+):
+    """
+    Given a SQLAlchemy SELECT query, apply the aggregation specified by `function_name`
+    to `input_colum`, grouping by `group_by_column` and labelling the result as
+    `output_column`
+    """
+    if function_name == "exists":
+        aggregate_value = sqlalchemy.literal(True)
+    else:
+        function = getattr(sqlalchemy.func, function_name)
+        source_column = query.selected_columns[input_column]
+        aggregate_value = function(source_column)
+
+    query = query.with_only_columns(
+        [
+            query.selected_columns[group_by_column],
+            aggregate_value.label(output_column),
+        ]
+    )
+    return query.group_by(query.selected_columns[group_by_column])
+
+
+def get_joined_tables(select_query):
+    """
+    Given a SELECT query object return a list of all tables referenced
+    """
+    tables = []
+    from_exprs = list(select_query.get_final_froms())
+    while from_exprs:
+        next_expr = from_exprs.pop()
+        if isinstance(next_expr, sqlalchemy.sql.selectable.Join):
+            from_exprs.extend([next_expr.left, next_expr.right])
+        else:
+            tables.append(next_expr)
+    # The above algorithm produces tables in right to left order, but it makes
+    # more sense to return them as left to right
+    tables.reverse()
+    return tables
+
+
+def get_primary_table(select_query):
+    """
+    Return the left-most table referenced in the SELECT query
+    """
+    return get_joined_tables(select_query)[0]
+
+
+def include_joined_tables(select_query, tables, join_column):
+    """
+    Ensure that each table in `tables` is included in the join conditions for
+    `select_query`
+    """
+    current_tables = get_joined_tables(select_query)
+    for table in tables:
+        if table in current_tables:
+            continue
+        join = sqlalchemy.join(
+            select_query.get_final_froms()[0],
```

Comment:
I don't think it's 100% obvious to me either! This relates a bit to a comment I made elsewhere about the population table. But I think what's going on here is that the first table in query always represents the universe of patients we're interested in. Either because it's the "population table" in the big query we do at the end, or because it's the table we're adding filter conditions to.